src/gui/editorwidgets/qgstexteditwrapper.cpp

```diff
@@ -177,6 +199,18 @@ void QgsTextEditWrapper::showIndeterminateState()
     mLineEdit->blockSignals( false );
 }
 
+void QgsTextEditWrapper::setFeature( const QgsFeature &feature )
+{
+  // Do nothing if the value has not changed
+  if ( mInvalidJSON )
+    mForm->displayWarning( tr( "Your JSON is invalid and will be reverted back to the last valid edit or the original data" ) );
```

Comment:
I'm not native English speaker but I think the user will not expect that it "will be" reverted. He expects that it has already been reverted (when he changed the feature). So I would suggest this.

```suggestion
    mForm->displayWarning( tr( "Your JSON was invalid and is reverted back to the last valid edit or the original data" ) );
```

Reply:
Have changed the wording slightly. I believe the other commits fix both issues. I think the freezing was also a validation problem.